spec/user_spec.rb

```diff
@@ -66,18 +171,278 @@
       expect(@subject.keys).to include("link_color:time_of_assignment")
     end
 
+    describe "when the user is in the experiment with version number" do
+      let(:user_keys) { {
+        'link_color:1' => 'blue',
+        'link_color:1:finished' => true,
+        'link_color:1:time_of_assignment' => Time.now.to_s,
+      } }
+
+      it 'keeps keys if the experiment has no winner and has started' do
+        allow(Split::ExperimentCatalog).to receive(:find).with('link_color').and_return(experiment)
+        allow(experiment).to receive(:has_winner?).and_return(false)
+        allow(experiment).to receive(:start_time).and_return(Date.today)
+
+        @subject.cleanup_old_experiments!
+
+        expect(@subject.keys).to include("link_color:1")
+        expect(@subject.keys).to include("link_color:1:finished")
+        expect(@subject.keys).to include("link_color:1:time_of_assignment")
+      end
+    end
+
     context 'when already cleaned up' do
       before do
         @subject.cleanup_old_experiments!
       end
 
       it 'does not clean up again' do
-        expect(@subject).to_not receive(:keys_without_finished_and_time_of_assignment)
+        expect(@subject).to_not receive(:experiment_keys)
         @subject.cleanup_old_experiments!
       end
     end
   end
 
+  context "#max_experiments_reached?" do
+    context "when multiple experiments are not allowed" do
+      before do
+        Split.configuration.allow_multiple_experiments = false
+      end
+
+      context "when user is not in any experiment" do
+        let(:user_keys) { {
+        } }
+
+        it "max not reached when checking against same experiment" do
+          expect(@subject.max_experiments_reached?("link_color")).to be_falsey
+        end
+      end
+
+      context "when user is already in an experiment" do
+        let(:user_keys) { {
+          'link_color' => 'blue',
+          'link_color:finished' => true,
+          'link_color:time_of_assignment' => Time.now.to_s,
+          'link_color:eligibility' => "ELIGIBLE",
+        } }
+
+        it "max reached when checking against another experiment" do
+          expect(@subject.max_experiments_reached?("link")).to be_truthy
+        end
+
+        it "max reached when checking against another experiment contain current experiment as substring" do
+          expect(@subject.max_experiments_reached?("link_color_v2")).to be_truthy
+        end
+
+        it "max not reached when checking against same experiment" do
+          expect(@subject.max_experiments_reached?("link_color")).to be_falsey
+        end
+
+        it "max reached when checking against same experiment but different version" do
+          expect(@subject.max_experiments_reached?("link_color:1")).to be_truthy
+        end
```

Comment:
I think the spec is ok, though i'm thinking there shouldn't be a scenario in actual code where we are checking `max_experiments_reached?` when a user has an experiment key of `link_color` and we're validating `link_color:1`. Where `max_experiments_reached?` is called in code ~ forcing the return of control shouldn't happen when `link_color` and `link_color:1` should be treated as the same experiment if `retain_user_alternatives_after_reset == true` (I realize there's an existing/unnoticed bug in https://github.com/clio/split/blob/master/lib/split/trial.rb#L159 where it should probably be called with the user's alternative key?). 

When `retain_user_alternatives_after_reset == false`, the user's previous versions should have been deleted (cleanup_old_versions) such that `link_color` will never exist to compare to `link_color:1`